### ISSUE TITLE
Use `p2p` Disconnects in `eth` handshake

### ISSUE THREAD

**Issue by Rjected** [paradigmxyz/reth#1492]
### Describe the feature

Currently we just terminate the stream when receiving a message, and do not send a `Disconnect` message when disconnecting. The `EthStream` treats the underlying stream it wraps as a generic `Stream` and `Sink`, so we will need to add an additional bound that is implemented as a no-op by standard stream types like `TcpStream`.

Something like the following:
```rust

/// This trait is meant to allow higher level protocols like `eth` to disconnect from a peer, using
/// lower-level disconnect functions (such as those that exist in the `p2p` protocol) if the
/// underlying stream supports it.
#[async_trait::async_trait]
pub trait CanDisconnect<T>: Sink<T> + Unpin + Sized {
    /// Disconnects from the underlying stream, using a [`DisconnectReason`] as disconnect
    /// information if the stream implements a protocol that can carry the additional disconnect
    /// metadata.
    async fn disconnect(
        &mut self,
        reason: DisconnectReason,
    ) -> Result<(), <Self as Sink<T>>::Error>;
}
```

This will then be implemented by `P2PStream`, so an `EthStream<P2PStream<S>>` can access this method on its `inner` field.

### Additional context

Hive `devp2p` tests depend on this functionality, and reth currently fails due to this:
https://github.com/paradigmxyz/reth/actions/runs/4237113733/jobs/7362781160#step:7:34

This is where hive checks for a disconnect in the `devp2p` tests:
https://github.com/ethereum/go-ethereum/blob/fe01a2f63b8591d8226742726d8d6aaad4cd981e/cmd/devp2p/internal/ethtest/helpers.go#L511

**Comment by mattsse** [comment]
Closed by #1494 
?

**Comment by Rjected** [comment]
> Closed by #1494
?

yep, forgot to add the fixes line
